### PR TITLE
Uart

### PR DESCRIPTION
**Pull request by csagonero**
Add uart hal with interrupt management

### CODE REVIEW

File: stm32f0_hal/src/uart.rs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+                .disabled()
+                .ctsie()
+                .disabled()
+                .ovrdis()
+                .disabled()
+        });
+        // Configure UART : baudrate
+        uart1.brr.write(|w| {
+            w.div_fraction().bits(
+                (48000000 / (baudrate / 2)) as u8 & 0x0F,
```
Comment (pierre-rouanet) [inline stm32f0_hal/src/uart.rs]: Use constant for frequency here instead.

File: stm32f0_hal/src/uart.rs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            )
+        });
+        uart1.brr.write(|w| {
+            w.div_mantissa().bits(
+                ((48000000 / (baudrate / 2)) >> 4) as u16,
```
Comment (pierre-rouanet) [inline stm32f0_hal/src/uart.rs]: Same here.

File: stm32f0_hal/src/uart.rs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        uart1.tdr.write(|w| w.tdr().bits(byte));
+    })
+}
+
+pub fn transmit_complete() -> bool {
+    cortex_m::interrupt::free(|cs| {
+        let uart1 = UART1.borrow(cs);
+        if uart1.isr.read().tc().bit_is_set() {
+            uart1.icr.write(|w| w.tccf().clear_bit());
+            return true;
```
Comment (pierre-rouanet) [inline stm32f0_hal/src/uart.rs]: simply *true* would be more rust-like?

File: stm32f0_hal/src/uart.rs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        } else {
+            return false;
```
Comment (pierre-rouanet) [inline stm32f0_hal/src/uart.rs]: Same here.

File: stm32f0_hal/src/uart.rs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+pub enum PARITY {
+    /// No Parity
+    None,
+    /// Even Parity
+    Even,
+    /// Odd Parity
+    Odd,
+}
+
+pub fn init(baudrate: u32, nbits: NBITS, nbstopbits: STOPBITS, parity: PARITY) {
```
Comment (pierre-rouanet) [inline stm32f0_hal/src/uart.rs]: Could we directly specify a reception callback here?

File: stm32f0_hal/src/uart.rs [hunk 6]
```diff
@@ -0,0 +1,176 @@
+use cortex_m;
+use stm32f0x2::{NVIC, RCC, GPIOA, USART1 as UART1};
+use stm32f0x2::interrupt::*;
+
+static mut DATA_UART1: u16 = 0;
+
+pub enum NBITS {
```
Comment (pierre-rouanet) [inline stm32f0_hal/src/uart.rs]: CamelCase? NBits?

File: stm32f0_hal/src/uart.rs [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    _8bits,
+    _9bits,
+}
+
+pub enum STOPBITS {
```
Comment (pierre-rouanet) [inline stm32f0_hal/src/uart.rs]: StopBits

File: stm32f0_hal/src/uart.rs [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    _0b5,
+    _1b,
+    _1b5,
+    _2b,
+}
+
+pub enum PARITY {
```
Comment (pierre-rouanet) [inline stm32f0_hal/src/uart.rs]: Parity?

File: stm32f0_hal/src/uart.rs [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+            |_, w| w.linen().disabled().clken().disabled(),
+        );
+        // UART1 enabled
+        uart1.cr1.modify(|_, w| w.ue().enabled());
+        nvic.enable(Interrupt::USART1);
+        nvic.clear_pending(Interrupt::USART1);
+    })
+}
+
+pub fn send(byte: u16) {
```
Comment (pierre-rouanet) [inline stm32f0_hal/src/uart.rs]: Shall we use an u8 for the moment?

File: stm32f0_hal/src/uart.rs (hunk 6 above)
Comment (csagonero) [reply stm32f0_hal/src/uart.rs]: Ok

**Review by pierre-rouanet — APPROVED** [review]

**Pull request merged**